doc/design_docs/strings.md

```diff
@@ -0,0 +1,321 @@
+    the correct length.
+
+    `FixString` is very close to a notional `Blob` type or the current
+    `UInt:8[]` type, except for differences in text format.
+
+2.  `ZString`, a terminated string.  A `ZString` with no arguments uses a null
+    byte (`'\0'`) as the terminator.  An optional argument can be used to
+    specify the terminator -- a `ZString(36)`, for example, would be terminated
+    by `$`.  When reading, the value returned is all bytes up to, but not
+    including, the first terminator byte.  When writing, for compatibility, the
+    entire field should be written, using the terminator value for padding if
+    there is extra space.  A second optional parameter can be used to specify
+    that the terminator is not required: `ZString(0, false)` can fill the
+    underlying field with no terminator.
+
+    `CouldWriteValue()` should return `true` if the value is no longer than the
+    field and the value does not *contain* any instances of the terminator
+    byte.
+
+3.  `PaddedString`, a padded string.  A `PaddedString` with no arguments uses
+    space (`' '`, 32) as the padding value.  An optional argument can be used to
+    specify the padding -- a `PaddedString(0)`, for example, would be padded
+    with null bytes.  When reading, the end of the string is discovered by
+    walking *backwards* from the end until a non-padding byte is found, then
+    returning all bytes from the start of the string to the end.  When writing,
+    any excess bytes will be filled with the padding value.
+
+    Although, technically, "at least one byte of padding" could be enforced by
+    making the `PaddedString` one byte shorter and following it with a one-byte
+    field whose value *must* be the padding byte, for convenience `PaddedString`
+    should take a second optional parameter to specify that the terminator *is*
+    required: `PaddedString(32, true)` must have at least one space at the end.
+
+    `CouldWriteValue()` should return `true` if the value is no longer than the
+    field and the value does not *end with* the padding byte.
+
+
+### String Constants
+
+String constants (used in constructs such as `[requires: this == "abcd"]`) may
+take two forms:
+
+1.  `"A quoted string using C-style escapes like \n"`
+
+    In addition to standard C89 escapes (as interpreted by an ASCII Unix
+    compiler):
+
+    *   `\0` => 0
+    *   `\a` => 7
+    *   `\b` => 8
+    *   `\t` => 9
+    *   `\n` => 10
+    *   `\v` => 11
+    *   `\f` => 12
+    *   `\r` => 13
+    *   `\"` => 34
+    *   `\'` => 39
+    *   `\?` => 63 (part of the C standard, but rarely used)
+    *   `\\` => 92
+    *   <code>\x*hh*</code> => 0x*hh*
+
+    The following non-C-standard escapes should be allowed:
+
+    *   `\e` => 27 (not actually standard, but common)
+    *   <code>\d*nnn*</code> => *nnn*
+    *   <code>\x{*hh*}</code> => 0x*hh*
+    *   <code>\d{*nnn*}</code> => *nnn*
```

Comment:
Do the brace-delimited forms explicitly require 2 or 3 characters? If multibyte characters are allowed in the future then presumably there'd be some kind of syntax like \x{hhhh} for a two-byte value since \xhhhh would be ambiguous.

Even with single byte values, I think \x1 and \d10 are not not valid and should instead be \x01 and \d010. But are \x{1} and \d{10} valid?

Reply:
Updated the section on brace-delimited forms to be clearer.

Added text that `\x1` or `\d10` should be disallowed.  I'm not 100% convinced that it's the right choice, but a basic design principle I've always followed in Emboss is that it's much much easier to start strict and gradually allow more things than to start loose and try to disallow things after the fact.

I also added some text to the effect that unknown escape sequences should not be allowed.